operator/pkg/nodeipam/nodesvclb.go

```diff
@@ -0,0 +1,242 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package nodeipam
+
+import (
+	"context"
+	"slices"
+	"sort"
+
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	discoveryv1 "k8s.io/api/discovery/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/selection"
+	"k8s.io/apimachinery/pkg/util/sets"
+	utilsnet "k8s.io/utils/net"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	controllerruntime "github.com/cilium/cilium/operator/pkg/controller-runtime"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+)
+
+var nodeSvcLBClass = "io.cilium/node"
+
+type nodeSvcLBReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+	Logger logrus.FieldLogger
+}
+
+func newNodeSvcLBReconciler(mgr ctrl.Manager, logger logrus.FieldLogger) *nodeSvcLBReconciler {
+	return &nodeSvcLBReconciler{
+		Client: mgr.GetClient(),
+		Scheme: mgr.GetScheme(),
+		Logger: logger,
+	}
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *nodeSvcLBReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	filter := func(obj client.Object) bool {
+		svc, ok := obj.(*corev1.Service)
+		return ok && r.isServiceSupported(svc)
+	}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		// Watch for changes to Services supported by the controller
+		For(&corev1.Service{}, builder.WithPredicates(predicate.NewPredicateFuncs(filter))).
+		// Watch for changes to EndpointSlices
+		Watches(&discoveryv1.EndpointSlice{}, r.enqueueRequestForEndpointSlice()).
```

Comment:
If you are interested in endpoints shouldn't you also be watching for [discoveryv1.Endpoint](https://github.com/kubernetes/api/blob/v0.29.0/discovery/v1/types.go#L77) objects?

Reply:
Right, but wouldn't you want to support Endpoint objects alongside EndpointSlices? Perhaps I am misremembering how this works, but isn't the EndpointSlice a newer API but the older Endpoint objects can also still be used to map services to pods?

Reply:
Yes it's the newer API indeed but the kube controller will mirror the Endpoints to EndpointSlice, see this: https://kubernetes.io/docs/concepts/services-networking/endpoint-slices/#endpointslice-mirroring. And it's starting to be pretty old now (stable since Kubernetes 1.21 and beta/enabled since 1.17 IIUC) .

Technically I could add some code to select if we would need to use an Endpoint or an EndpointSlice but I would be very surprised if anyone has a use case for that :thinking:.

Reply:
Actually given Cilium support matrix of kubernetes is 1.19, maybe I could add a check with a warning to prevent running the controller in version older than 1.21?

Given that it's an optional features and that it should (hopefully) land in the next release of Cilium in like give or take 6 months making kubernetes 1.19 and 1.20 even more old I am not sure it's really worth it to add Endpoint compatibility... But I think that a check with a warning or error perhaps is good middle ground, WDYT?

Reply:
> Yes it's the newer API indeed but the kube controller will mirror the Endpoints to EndpointSlice

I didn't know about this. Thanks!

> Actually given Cilium support matrix of kubernetes is 1.19 minimum, maybe I could add a check with a warning to prevent running the controller in version older than 1.21?

Updating the support matrix has come up in the past, I don't think there are many users running <1.21. So I would let this slide.

Thanks for checking!